BUG: fix moments method to support arrays and list

tirthasheshpatel opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
fixes #12192

#### What does this implement/fix?
<!--Please explain your changes.-->
This pr adds list and array support for ``moment`` method of every distribution.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1203,7 +1204,7 @@ def moment(self, n, *args, **kwds):
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        if loc == 0:
+        if np.all(loc == 0):
```
I haven't read the surrounding code, but should this do different things for the elements that are zero vs the ones that aren't zero?

tirthasheshpatel replied:
Oh, true! sorry for that. I will take a deeper look and add tests soon. Thanks for the review

tirthasheshpatel replied:
@mdhaber I think I have done it right this time. I can't test it as I am unable to run tests. When I try to run tests using pytest on Linux, I get this error: ``AttributeError: module 'scipy.interpolate.dfitpack' has no attribute 'types'`` at line 38 of ``scipy/interpolate/_fitpack_impl.py``. I built it twice from source and it builds too. Is there anything I am missing?

mdhaber replied:
What does `git status` say?
Are you using `runtests.py` from the command line or some other way of running tests?
What instructions did you use to build SciPy?

tirthasheshpatel replied:
Aha! Using ``runtests.py`` did work! Is there a way to run specific tests only?

tirthasheshpatel replied:
Thanks! ``runtests.py`` did work!!

mdhaber commented:
There is a section in the [contributor guide](http://scipy.github.io/devdocs/dev/contributor/contributor_toc.html) that answers your question.

pvanmulbregt commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1186,8 +1186,14 @@ def moment(self, n, *args, **kwds):
         args, loc, scale = self._parse_args(*args, **kwds)
-        if not (self._argcheck(*args) and (scale > 0)):
-            return nan
+        arrs = tuple(map(np.asarray, [*args, loc, scale]))
```
This first assignment appears to be unused?

tirthasheshpatel replied:
`*args`? I think I did it because `loc` and `scale` parameters need to be of the same shape as inputs (that are present in `args`) otherwise the output isn't consistent with input shapes...

pvanmulbregt replied:
The first line is not needed.
```
arrs = tuple(map(np.asarray, [*args, loc, scale]))
arrs = tuple(map(np.atleast_1d, [*args, loc, scale]))
```

Doing this causes the return value to always be an array if I'm understanding the code correctly.

ev-br replied:
atleast_1d receives a list of arrays already, so the tuple(map(...)) dance can be avoided.

tirthasheshpatel replied:
Done. Thanks for the review!

mdhaber replied:
Do you still need the `arrs = tuple(map(np.asarray, [*args, loc, scale]))` even? 
`arrs = np.broadcast_arrays(*(*args, loc, scale))` accepts inputs that are not already arrays.

pvanmulbregt commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1203,16 +1209,17 @@ def moment(self, n, *args, **kwds):
-                result += comb(n, k, exact=True)*(fac**k) * valk
-            result += fac**n * val
-            return result * loc**n
+        cond1 = (loc == 0)
```
This appears to do all the additional work even for `loc=0`.  You should be able to use `argsreduce` to restrict to the values that need computing. E.g. 
```
cond = (loc == loc) & (scale == scale)
if np.any(cond):
  goodargs = argsreduce(cond, loc, scale)
  ...
  place(result, cond, ...)
```

tirthasheshpatel replied:
I have tried to use `_lazywhere` to fill `nan` and circumvent redundant calculations. Is it ok or should I try for this fix?

mdhaber replied:
I think it still has to do some sort of calculation there, though. You use `_lazywhere` to fill `fac`with `nan`s, but as far as I can tell, that doesn't really circumvent redundant calculations, right?  All sorts of operations are done on `fac`. For the `nan`s, these operations aren't doing anything useful, right? I do think it would be a good idea to try implementing this suggestion.

mdhaber commented:
I forgot about this one, @tirthasheshpatel. I can start reviewing this one too, but it looks like there are some comments already.

tirthasheshpatel commented:
@mdhaber Sorry for stalling this. I will try to resume my work on this in a couple of days!

mdhaber commented:
Hi @tirthasheshpatel, just thought I should check in. Are you still interested in continuing with this and gh-12585? We really appreciate it!

tirthasheshpatel commented:
@mdhaber I am very sorry to delay my work on this PR. I won't be able to continue my work on this while working on #12585. I will open a PR for #12585 as soon as I get it working. I hope you don't mind. Again, sorry for wasting so much time!

mdhaber commented:
@tirthasheshpatel no problem; we'll get back to this after merging gh-12585

mdhaber commented:
@tirthasheshpatel I resolved the merge conflict. Please check that the changes from the first six commits and the changes from all commits are basically identical (I think I accidentally added a PEP8 newline). 

I think gh-13103 is actually going to take a little while to figure out. In the meantime, if you'd like to come back to this, I'd be happy to review.

tirthasheshpatel commented:
Thanks for resolving the conflicts, @mdhaber! I have pushed some new changes that should work fine but some redundant calculations still exists. I will try to figure out a way to work around those too. Meanwhile, would appreciae a review.

mdhaber commented:
Sure I started taking a look. Can you add those tests?

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1240,8 +1240,13 @@ def moment(self, n, *args, **kwds):
+        arrs = np.broadcast_arrays(*arrs)
+        args = arrs[:-2]
+        loc, scale = arrs[-2], arrs[-1]
+        cond0 = (self._argcheck(*args) and scale > 0)
```
```suggestion
        cond0 = (self._argcheck(*args) & scale > 0)
```
Currently isn't working when there are shapes:
```
import numpy as np
from scipy.stats import chi as dist
m = dist.moment(n=3, df = 1, loc=[1, 2, 3], scale=1)
```

mdhaber replied:
Otherwise, this stuff up top looks good to me.

tirthasheshpatel replied:
I guess this is fixed now.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1257,16 +1262,18 @@ def moment(self, n, *args, **kwds):
 
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        if loc == 0:
-            return scale**n * val
-        else:
-            result = 0
-            fac = float(scale) / float(loc)
-            for k in range(n):
-                valk = _moment_from_stats(k, mu, mu2, g1, g2, self._munp, args)
-                result += comb(n, k, exact=True)*(fac**k) * valk
-            result += fac**n * val
-            return result * loc**n
```
I don't think there's anything wrong here, but as the one who might merge this, here's my preference:

Since this is a really fundamental function, I'd like to see how little we can change these lines. It would be great if you could even keep the original variable name `result`. If you implement @pvanmulbregt's suggestion to keep only the elements that meet both `cond`itions, I think you can keep these lines almost exactly the same as they are. They're pretty close right now, but please even avoid the refactoring (e.g. keep the `+=` as they are, etc.)

tirthasheshpatel replied:
> creating a result array of zeros only large enough to hold the results for the elements that meet both conditions - the ones that need to go through the iterative procedure

done.

> minimal changes to the iterative procedure

done.

> reassembling the output - putting in nans where cond0 wasn't met and putting scale**n * val where cond1 wasn't met

done.

I think that the current implementation should work fine for all array inputs too.

tirthasheshpatel commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1252,21 +1261,25 @@ def moment(self, n, *args, **kwds):
+            result[cond0] += comb(n, k, exact=True)*fac**k * valk
+        result[cond0] += fac**n * val
+        result[cond0] *= goodloc**n
+        res = goodscale**n * val
```
Apparently, this implementation fails when `(loc==0) & (scale <= 0)`. For instance, the following calls fail:

```python
from scipy.stats import norm

norm.moment(2, 0, -4)
norm.moment(2, [1, 0, 2], [3, -4, -5])
```

Will try to address this and add some tests in the next commit

tirthasheshpatel commented:
I have added some tests and removed, I think, all the redundant calculations. This is ready for review @mdhaber, @pvanmulbregt  if anyone of you wanna have a look.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1252,21 +1261,25 @@ def moment(self, n, *args, **kwds):
-            result += fac**n * val
-            return result * loc**n
+        fac = _lazywhere(cond1_, (goodloc, goodscale),
+                         lambda loc, scale: scale / loc, self.badvalue)
```
Since we're still adding `nan`s here where `loc==0`, we still do operations on those elements below.  Ideally, we wouldn't be touching the elements where `loc == 0` until we get to  `res = goodscale**n * val`. And ideally, that statement shouldn't touch the elements where `loc != 0`. I think the solution would be to do another `argsreduce` after you get `val` for all the `goodargs`.

I'll submit a PR to your branch so you can see what I mean.

tirthasheshpatel replied:
Thanks! I merged it from my fork.

mdhaber commented:
Oh, OK. Cool. Well, I'm not really an independent reviewer anymore, so we might want to get @pvanmulbregt's opinion on the new approach.

pvanmulbregt commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -654,3 +654,84 @@ def test_methods_with_lists(method, distname, args):
+    # array loc and scalar scale
+    vals0 = stats.norm.moment(n=1, loc=np.array([1, 2, 3]), scale=1)
+    expected0 = np.array([1., 2., 3.])
+    npt.assert_allclose(vals0, expected0, rtol=1e-20)
```
Why rtol=1e-20? Anything less than np.finfo(float).eps ~1e-16 is requesting equality.

tirthasheshpatel replied:
Oh, I didn't know that. I will change it to equality then.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1277,7 +1277,7 @@ def moment(self, n, *args, **kwds):
                 mu, mu2, g1, g2 = argsreduce(loc != 0, mu, mu2, g1, g2)
             args = argsreduce(loc != 0, *shapes, loc, scale, val)
-            shapes, loc, scale, val = args[:-2], args[-3], args[-2], args[-1]
+            shapes, loc, scale, val = args[:-3], args[-3], args[-2], args[-1]
```
Thanks. Typo.

mdhaber commented:
Something to check before merging - do scalar inputs to discrete distribution `moment` method enter `_argcheck` as scalars or arrays? Currently, they enter as scalars, but they should enter it as arrays (as for other methods).

tirthasheshpatel commented:
> do scalar inputs to discrete distribution moment method enter _argcheck as scalars or arrays?

They enter as arrays. They are converted to arrays in the `np.broadcast_arrays` call.

tirthasheshpatel commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -662,3 +662,89 @@ def test_burr_fisk_moment_gh13234_regression():
+    expected9 = np.array([3.59576912, np.nan, np.nan])
+    npt.assert_allclose(vals9, expected9, rtol=1e-8)
+
+    # (n > 4), all(loc != 0), and all(scale != 0)
+    vals10 = stats.norm.moment(5, [1., 2.], [1., 2.])
+    expected10 = np.array([26., 832.])
+    npt.assert_equal(vals10, expected10)
```
@mdhaber I just noticed that the failing test is due to a small numerical difference in the calculated result. So, we will need to change this to `assert_allclose` from `np.assert_equal`.

mdhaber commented:
@pvanmulbregt Would you be able to come back to this? I think it's looking pretty good, but since I ended up contributing a bit I'd prefer to have another reviewer.

mdhaber commented:
@tylerjereddy I can take another look at this today and merge if need be, but I was hoping to have an independent review since I contributed some of it. I won't put it on the milestone, but I think it's important because currently the `moment` method of stats distributions says it supports array inputs (like every other method) for `loc`, `scale`, and shape parameters, but doesn't.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1283,21 +1291,37 @@ def moment(self, n, *args, **kwds):
+            place(result, i1, res1)
+
+        if i2.any():
+            if mu is not None:
```
I'm concerned about this line. There are some distributions for which `mu` might not be none but for which the others (`mu2`, etc...) will be `None`.

mdhaber replied:
Looks like the test failure might be caused by that. I thought this was ready, but I guess not : /

tirthasheshpatel replied:
> Looks like the test failure might be related because `dweibull` is one of those distributions. I thought this was ready, but I guess not : /

I have tried to resolve this in the latest commit. Do you think the changes are OK? If yes, is it still possible to get this merged for 1.7?

tylerjereddy replied:
Since it is labelled a bug fix it may need to get backported before the rc or during the RC phase, which is fine if it really is important enough to do that.

tylerjereddy replied:
But probably not before I branch, since I'm about to do that.

tirthasheshpatel replied:
Thanks @tylerjereddy for the replies! I wouldn't say it's top priority but it would be nice to have it in 1.7 since it has been up for a while now and fixes quite a bit to avoid workarounds in other PRs. Though, if it means a lot of work has to be done from your side, it would be fine to merge this later too.

tylerjereddy replied:
Shouldn't be a problem to put it in, I just mean it may get delayed until after I branch and then get added in with other bug fixes shortly after branching.

tirthasheshpatel replied:
> Shouldn't be a problem to put it in, I just mean it may get delayed until after I branch and then get added in with other bug fixes shortly after branching.

Oh, OK. That's fine! Thanks for the clarification.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1283,21 +1291,37 @@ def moment(self, n, *args, **kwds):
+            res2 *= loc**n
+            place(result, i2, res2)
+
+        if result.ndim == 0:
+            return result.item()
+        return result
```
```suggestion
        return result[()]
```
I think we decided in another conversation that this was preferable.

tirthasheshpatel replied:
I had to revert this as a test required the returned moment to be a Python float when scalar.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1283,21 +1291,45 @@ def moment(self, n, *args, **kwds):
+
+        if i2.any():
+            mom = [mu, mu2, g1, g2]
+            arrs = [i for i in mom if i is not None]
```
Was the problem that we assumed if `mu` was `None` then the rest would be `None`?

Could you explain what is going on here?

tirthasheshpatel replied:
> Was the problem that we assumed if mu was None then the rest would be None?

Yes, I think so.

> Could you explain what is going on here?

I am creating a list of moments that are not `None` and calling `argsreduce` to remove the invalid entries. But as we don't know which of them would be `None` and which wouldn't, I have also created a list of indices recording the index of the moments that are not `None`. Once `argsreduce` returns a list of reduced moments, I just fill in the `mom` list with the moments that have been reduced (using the index list).

I didn't think much about a better approach to get this done quickly. If this is acceptable for now, it should work just fine and we can get back to it later and do something that is clearer. Does that sound good?

tirthasheshpatel replied:
> Would you reach out to some other core developers who might be interested?

Sure.

> I'll take a look at some of those PRs you submitted for adding alternative to mstats hypothesis tests.

Thanks!

tirthasheshpatel commented:
@tupui @WarrenWeckesser Can any of you help review this, please? I and @mdhaber have been working on this and need an independent reviewer to verify the approach. If one of you has time to review, it would be very helpful! Thanks!

tupui reviewed (commented):
Sorry for the delay I was away. Just some quick comment for now, but I will try to check the logic ASAP.

tirthasheshpatel commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1270,9 +1270,17 @@ def moment(self, n, *args, **kwds):
-            return nan
+        shapes, loc, scale = self._parse_args(*args, **kwds)
+        args = np.broadcast_arrays(*(*shapes, loc, scale))
+        shapes, loc, scale = args[:-2], args[-2], args[-1]
```
Oh, wow! I didn't know Python allows this syntax!

tirthasheshpatel commented:
Thanks very much for the review, @tupui! I have applied all your suggestion now.

tupui commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1283,21 +1291,45 @@ def moment(self, n, *args, **kwds):
+                    j+=1
+            mu, mu2, g1, g2 = mom
+            args = argsreduce(loc != 0, *shapes, loc, scale, val)
+            shapes, loc, scale, val = args[:-3], args[-3], args[-2], args[-1]
```
I forgot this one
```suggestion
            *shapes, loc, scale, val = args
```

tupui reviewed (approved):
I had a bit of time to go over again. LGTM, the logic looks correct just a minor thing I missed last time, and you added quite some tests (I tested this manually and it checks out in terms of shape and values). Thanks for the great work. @mdhaber feel free to merge (not sure if you want to squash or merge, you are more involved).

mdhaber commented:
@Zac-HD This (and other methods of SciPy stats distributions) could be a good example of where Hypothesis would be useful in SciPy.

We'd like to check that all methods of all distributions are vectorized properly. This PR is just about the `moment` method, so let's take the `moment` method of the normal distribution as an example.
```python3
norm.moment(n, loc=loc, scale=scale)
```
 should behave the same as 
```python3
ref(n, loc=loc, scale=scale)
```
where 
```python3
ref = np.vectorize(norm.moment, otype=[float])
```
and `n` is an integer (the order of the moment), `loc` (the location parameter of the distribution) a real or array-like of reals, and `scale` (the scale parameter of the distribution) is a real or array-like of reals.

One minor complication is that some distributions have additional shape parameters, which have varying valid ranges, and we'd want to make sure that at least some combinations of inputs are valid. (When there is an invalid parameter, there will be a `nan` in the output array.)

Would you be able to help us set up something - just for the `moment` method, and we could probably extend it from there - like you did for NumPy in numpy/numpy#15189?

Zac-HD commented:
I'd be delighted to help out!  Here's a quick prototype:

```python
import numpy as np
import hypothesis.extra.numpy as npst
from hypothesis import given, strategies as st

from scipy.stats import norm

# A strategy to generate the shape of a one-dimensional array of length [0..7]
# Subsequent draws will use the same shape to ensure compatibility.  For fancier
# tricks see `npst.broadcastable_shapes()` or `mutually_broadcastable_shapes()`.
shared_shapes = st.shared(st.tuples(st.integers(0, 7)))


def scalar_or_array(**kwargs):
    return st.floats(**kwargs) | npst.arrays(
        dtype=float, shape=shared_shapes, elements=kwargs
    )


@given(
    # Our moment `n`, between one and four (inclusive)
    n=st.integers(1, 4),
    # Our locations and scale are each either a scalar (Python-native) float,
    # or an ndarray of floats.  If both are arrays, they will have the same shape.
    loc=scalar_or_array(min_value=0, max_value=1000),  # what should the bounds be?
    scale=scalar_or_array(min_value=1, max_value=1000),
)
def test_moments(n, loc, scale):
    got = norm.moment(n, loc=loc, scale=scale)
    ref = np.vectorize(norm.moment, otypes=[float])(n, loc=loc, scale=scale)

    assert type(got) == type(ref)  # fails; vectorize returns zero-dim arrays
    if isinstance(got, np.ndarray):
        assert got.shape == ref.shape
        assert got.dtype == ref.dtype
    np.testing.assert_allclose(got, ref)
```

On my machine with `scipy == 1.7.0`, this fails with the errors you'd expect of a version that doesn't support array inputs yet:

```python-traceback
________________________________ test_moments _________________________________
Traceback (most recent call last):
  File "C:\Users\Zac\Documents\GitHub\hypothesis\t.py", line 20, in test_moments
    # Our moment `n`, between one and four (inclusive)
  File "c:\users\zac\documents\github\hypothesis\hypothesis-python\src\hypothesis\core.py", line 1180, in wrapped_test
    raise the_error_hypothesis_found
hypothesis.errors.MultipleFailures: Hypothesis found 4 distinct failures.
--------------------------------- Hypothesis ----------------------------------
Falsifying example: test_moments(
    n=1, loc=array([0., 0.]), scale=1.0,
)
Traceback (most recent call last):
  File "C:\Users\Zac\Documents\GitHub\hypothesis\t.py", line 28, in test_moments
    got = norm.moment(n, loc=loc, scale=scale)
  File "c:\users\zac\anaconda3\envs\hydev\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1291, in moment
    if loc == 0:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

Falsifying example: test_moments(
    n=1, loc=0.0, scale=array([1., 1.]),
)
Traceback (most recent call last):
  File "C:\Users\Zac\Documents\GitHub\hypothesis\t.py", line 28, in test_moments
    got = norm.moment(n, loc=loc, scale=scale)
  File "c:\users\zac\anaconda3\envs\hydev\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1274, in moment
    if not (self._argcheck(*args) and (scale > 0)):
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

Falsifying example: test_moments(
    n=1, loc=array([], dtype=float64), scale=1.0,
)
Traceback (most recent call last):
  File "C:\Users\Zac\Documents\GitHub\hypothesis\t.py", line 28, in test_moments
    got = norm.moment(n, loc=loc, scale=scale)
  File "c:\users\zac\anaconda3\envs\hydev\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1295, in moment
    fac = float(scale) / float(loc)
TypeError: only size-1 arrays can be converted to Python scalars

Falsifying example: test_moments(
    n=1, loc=array([], dtype=float64), scale=array([], dtype=float64),
)
Traceback (most recent call last):
  File "C:\Users\Zac\Documents\GitHub\hypothesis\t.py", line 31, in test_moments
    assert type(got) == type(ref)
AssertionError: assert <class 'float'> == <class 'numpy.ndarray'>
 +  where <class 'float'> = type(nan)
 +  and   <class 'numpy.ndarray'> = type(array([], dtype=float64))
```

Hopefully this is a useful starting point?  Further elaboration will probably require deeper knowledge (or judgement) about scipy edge cases than I have, but I'd be up for a call or pairing session too if that would help 😁

tirthasheshpatel commented:
Hi @Zac-HD 

> I'd be delighted to help out!

Thanks!

>  Here's a quick prototype

I just had a quick look and it looks really good! I have never used hypothesis so I will read some docs today and can start cooking up my own version in some time. If you think you have something useful (which you possibly would before I learn to use hypothesis), feel free to just create a PR on my fork. I will merge it and we can then review it here.

> ```python
>     loc=scalar_or_array(min_value=0, max_value=1000),  # what should the bounds be?
>     scale=scalar_or_array(min_value=1, max_value=1000),
> ```

I would throw in some invalid values too to make it a more comprehensive test (). Something like:

```python
    loc=scalar_or_array(min_value=0, max_value=1000),  # what should the bounds be?
    scale=scalar_or_array(min_value=-1, max_value=1000),
```

tirthasheshpatel commented:
@Zac-HD @mdhaber I have added a basic test using hypothesis using @Zac-HD's prototype as a starting point. I removed the line:

```py
    assert type(got) == type(ref)  # fails; vectorize returns zero-dim arrays
```

because `np.vectorize` returns `np.ndarray` for scalar values while `stats.moment` returns python floats. We can also explicitly check for the condition when a scalar is expected and assert accordingly. The other thing I have changed is the range of `scale`. Instead of `(1, 1000)`, I have used `(-1, 1000)` so that sometimes invalid values are present. All tests pass locally. Let me know what you think.

tupui reviewed (commented):
I have nothing against adding `hypothesis` (although it adds yet another layer of complexity: pytests, asv, linters, now this), but might be worth an email in the mailing list to at least present it. Also, could you add it in https://scipy.github.io/devdocs/dev/toolchain.html? and there might be other places like the `environment.yml`.

mdhaber commented:
> might be worth an email in the mailing list to at least present it

Yes, that's [the plan](https://github.com/scipy/scipy/pull/12197#issuecomment-881100965). We just wanted to have a concrete example of how hypothesis can be used to improve tests.

mdhaber commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -719,3 +719,112 @@ def test_burr_fisk_moment_gh13234_regression():
+                          [np.nan, np.nan, 5.75994241],
+                          [np.nan, np.nan, 6.78730736]])
+    npt.assert_allclose(vals3, expected3, rtol=1e-8)
+    npt.assert_(vals3.shape == expected3.shape)
```
It looks like this was added but the original was not removed.

mdhaber commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -808,23 +810,54 @@ def test_broadcasting_in_moments_gh12192_regression():
+# A strategy to generate the shape of a one-dimensional array of length [0..7]
+# Subsequent draws will use the same shape to ensure compatibility.  For fancier
+# tricks see `npst.broadcastable_shapes()` or `mutually_broadcastable_shapes()`.
+shared_shapes = st.shared(st.tuples(st.integers(0, 7)))
```
Ideally, to show the advantage of Hypothesis, I think we want this test to go beyond what the tests above can do - show that this works for any distribution and any broadcastable `loc`, `scale`, and shape parameters. (I think it would also be good to nudge hypothesis toward occasionally testing with invalid parameters. I'm not sure how Hypothesis chooses values under the hood, but if we give it such a wide range for scale, is it likely to include in invalid value < 0 in one of the tests?). @tirthasheshpatel @Zac-HD if you are willing to go a little deeper into this, terrific; if not, I understand and I can get to this at some point.

Zac-HD replied:
I'd recommend having separate tests for always-valid and might-be-invalid inputs; otherwise you eventually and without realising write a test which never checks the valid case :sweat_smile:

For a deeper test, you're going to want [the `mutually_broadcastable_shapes()` strategy](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.mutually_broadcastable_shapes) - there's [an example in Numpy for `np.clip()` here](https://github.com/Zac-HD/numpy/blob/4b62c91878386b0a1c837aeb522ce3e2f9d62201/numpy/core/tests/test_numeric.py#L2337-L2384).  Where signatures are consistent you could use `@pytest.mark.parametrize` to supply the moment function - using that to supply some arguments and `@hypothesis.given()` for others works fine - but where signatures differ it's probably easier to write separate tests.

(This is getting pretty far from the original intent to fix a bug though, and I've seen projects take quite a while to settle on using Hypothesis.  Might be worth merging this PR, and then tackling the test changes in a new one?)

tirthasheshpatel replied:
Thanks for the feedback @Zac-HD and @mdhaber! I agree we should use Hypothesis to its fullest and add a deeper test. I have some time this and next week so I can go a little deeper and add a more comprehensive test.

> This is getting pretty far from the original intent to fix a bug though, and I've seen projects take quite a while to settle on using Hypothesis. Might be worth merging this PR, and then tackling the test changes in a new one?

I also think this would be better. Though I will try to push something or at least confirm locally using Hypothesis that the bug-fix works. I can then propose a PR and we can discuss Hypothesis in detail there. What do you think, @mdhaber?

mdhaber replied:
Yes, local test is fine. This PR has been tricky to get right, so a more comprehensive test would give additional confidence that it's ready to be merged. 
And yes, it is actually preferable for the test to go in a separate PR.
